generator/internal/sidekick/command.go

```diff
@@ -0,0 +1,256 @@
+	c.flags.BoolVar(p, name, value, usage)
+	return c
+}
+
+func (c *Command) AddFlagString(p *string, name string, value string, usage string) *Command {
+	c.flags.StringVar(p, name, value, usage)
+	return c
+}
+
+func (c *Command) AddFlagFunc(name string, usage string, fn func(string) error) *Command {
+	c.flags.Func(name, usage, fn)
+	return c
+}
+
+// Flags returns all flags added to this command, as well as its parent hierarchy
+func (c *Command) Flags() []*flag.Flag {
+	var flags = []*flag.Flag{}
+	c.VisitAllFlags(func(f *flag.Flag) {
+		flags = append(flags, f)
+	})
+	return flags
+}
+
+// VisitAllFlags visits all flags in the command, including those of its parent hierarchy
+func (c *Command) VisitAllFlags(fn func(f *flag.Flag)) {
+	c.flags.VisitAll(fn)
+	if c.parent != nil {
+		c.parent.VisitAllFlags(fn)
+	}
+}
+
+func (c *Command) Run(rootConfig *Config, cmdLine *CommandLine) error {
+	if !c.Runnable() {
+		return fmt.Errorf("command sidekick %s is not runnable", c.LongName())
+	}
+
+	return c.action(rootConfig, cmdLine)
+}
+
+func (c *Command) ParseCmdLine(args []string) (*CommandLine, error) {
+	if c.parent != nil {
+		c.parent.VisitAllFlags(func(f *flag.Flag) {
+			c.flags.Var(f.Value, f.Name, f.Usage)
+		})
+	}
+
+	if err := c.flags.Parse(args); err != nil {
+		return nil, err
+	}
+
+	return &CommandLine{
+		Command:             args,
+		ProjectRoot:         flagProjectRoot,
+		SpecificationFormat: format,
+		SpecificationSource: source,
+		ServiceConfig:       serviceConfig,
+		Source:              sourceOpts,
+		Language:            flagLanguage,
+		Output:              output,
+		Codec:               codecOpts,
+		DryRun:              dryrun,
+	}, nil
+}
+
+func NewCommand(
+	usageLine string,
+	shortDescription string,
+	longDescription string,
+	parent *Command,
+	action func(rootConfig *Config, cmdLine *CommandLine) error,
+) *Command {
+	if len(usageLine) == 0 {
+		panic("command usage line cannot be empty")
+	}
+	if !strings.HasPrefix(usageLine, "sidekick") {
+		panic(fmt.Sprintf("command usage line must start with sidekick, got: %s", usageLine))
+	}
+	if len(shortDescription) == 0 {
+		panic("command short description cannot be empty")
+	}
+	if parent == nil {
+		panic("command must have a parent")
+	}
+	c := newCommand(usageLine, shortDescription, parent)
+	c.longDescription = longDescription
+	c.action = action
+	return c
+}
+
+func newCommand(usageLine string, shortDescription string, parent *Command) *Command {
+	c := &Command{
+		usageLine:        usageLine,
+		altNames:         []string{},
+		shortDescription: shortDescription,
+		flags:            flag.NewFlagSet(usageLine, flag.ContinueOnError),
+		commands:         []*Command{},
+	}
+
+	// the only command that is valid without a parent is `sidekick`
+	if parent != nil {
+		parent.commands = append(parent.commands, c)
+	}
+	c.parent = parent
+
+	c.flags.Usage = func() {
+		_ = c.PrintUsage()
+	}
+	return c
+}
+
+var usageTemplate = `
```

Comment:
TBH, this didn't occur to me as I was doing it, but now that you mentioned I quite like the idea. I mostly just reverse engineered it from the Go implementation, but I think I have simplified the logic to a point that it may make more sense to just write the code for it rather than use a template (especially because we only have a single template). I'll give it a try and report back here.